bpo-40334: Fix error location upon parsing an invalid string literal

lysnikolaou opened the pull request:
When parsing a string with an invalid escape, the old parser used to
point to the beginning of the invalid string. This PR changes the new
parser to match that behavior, since it's currently pointing to the
end of the string (or to be more precise, to the beginning of the next
token).

Closes we-like-parsers/cpython#64.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-40334](https://bugs.python.org/issue40334) -->
https://bugs.python.org/issue40334
<!-- /issue-number -->

lysnikolaou committed:
bpo-40334: Fix error location upon parsing an invalid string literal

When parsing a string with an invalid escape, the old parser used to
point to the beginning of the invalid string. This PR changes the new
parser to match that behavior, since it's currently pointing to the
end of the string (or to be more precise, to the beginning of the next
token).

Closes we-like-parsers/cpython#64.

lysnikolaou committed:
Update test_cmd_line_script as well

pablogsal commented on Parser/pegen/pegen.c:
```diff
@@ -1977,7 +1977,7 @@ _PyPegen_concatenate_strings(Parser *p, asdl_seq *strings)
             goto error;
         }
 
-        if (_PyPegen_parsestr(p, this_str, &this_bytesmode, &this_rawmode, &s, &fstr, &fstrlen) != 0) {
```
If we pass the token, we can maybe get the value for `this_str` inside `_PyPegen_parsestr`

lysnikolaou replied:
Done!

pablogsal commented on Parser/pegen/parse_string.c:
```diff
@@ -12,20 +12,27 @@
+               since _PyPegen_raise_error always uses p->tokens[p->fill - 1] for the
+               error location. */
+            Token *old_end = p->tokens[p->fill - 1];
+            p->tokens[p->fill - 1] = t;
```
Instead, could we have a new  `Token*` attribute on the parser that (optionally) points to a token where the syntax error happens? Then, in the raise function, we take that token as a preference over `p->tokens[p->fill - 1]`. In this way, we don't need to modify the token string, which initially feels suboptimal (I have not thought about other side effects this may had).

lysnikolaou replied:
You mean in the `Parser` struct?

pablogsal replied:
> You mean in the `Parser` struct?

Yup

pablogsal replied:
Maybe `p->known_err_token` or something similar.

lysnikolaou replied:
Good idea! On it.

lysnikolaou replied:
Done.

pablogsal commented on Lib/test/test_cmd_line_script.py:
```diff
@@ -648,7 +648,7 @@ def test_syntaxerror_invalid_escape_sequence_multi_line(self):
                 stderr.splitlines()[-3:],
                 [   b'    foo = """\\q"""',
-                    b'                 ^',
+                    b'          ^',
```
Does this work also with the old parser?

lysnikolaou replied:
It does with `python -X oldparser -m test -v test_cmd_line_script`. I also checked on the REPL that the `SyntaxError` emitted has the same `col_offset`.

pablogsal reviewed (commented):
Nice! I left some comments and questions but initially, looks great :)

lysnikolaou committed:
Call PyBytes_AsString in _PyPegen_parsestr instead of _PyPegen_concatenate_strings

lysnikolaou committed:
Add a new field `known_err_token` to the `Parser` struct that has
precedence over `p->tokens[p->fill - 1]` for the `SyntaxError`
location and use it in `warn_invalid_escape_sequence`.

pablogsal reviewed (approved):
Great! LGTM

gvanrossum commented on Lib/test/test_string_literals.py:
```diff
@@ -118,8 +118,7 @@ def test_eval_str_invalid_escape(self):
         self.assertEqual(w[0].filename, '<string>')
-        if use_old_parser():
-            self.assertEqual(w[0].lineno, 1)
+        self.assertEqual(w[0].lineno, 1)
```
No assert for column?

lysnikolaou replied:
We cannot do this here, because `WarningMessage` objects do not hold column information. Added an assertion a bit further down, where a `SyntaxError` is thrown.

lysnikolaou committed:
Add assertion for column number

gvanrossum reviewed (approved):
How about the next test/chunk? Does it still pass with the old parser?
